Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean

```diff
@@ -0,0 +1,201 @@
+/-
+Copyright (c) 2023 Yaël Dillies, Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Bhavik Mehta
+-/
+import Mathlib.Algebra.DirectSum.AddChar
+import Mathlib.Analysis.Fourier.FiniteAbelian.Orthogonality
+import Mathlib.Analysis.SpecialFunctions.Complex.Circle
+import Mathlib.GroupTheory.FiniteAbelian
+
+/-!
+# Pontryagin duality for finite abelian groups
+
+This file proves the Pontryagin duality in case of finite abelian groups. This states that any
+finite abelian group is canonically isomorphic to its double dual (the space of complex-valued
+characters of its space of complex-valued characters).
+
+We first prove it for `ZMod n` and then extend to all finite abelian groups using the
+Structure Theorem.
+-/
+
+noncomputable section
+
+open Circle Finset Function Multiplicative
+open Fintype (card)
+open Real hiding exp
+open scoped BigOperators DirectSum
+
+variable {α : Type*} [AddCommGroup α] {n : ℕ} {a b : α}
+
+namespace AddChar
+variable (n : ℕ) [NeZero n]
+
+/-- Indexing of the complex characters of `ZMod n`. `AddChar.zmod n x` is the character sending `y`
+to `e ^ (2 * π * i * x * y / n)`. -/
+def zmod (x : ZMod n) : AddChar (ZMod n) Circle :=
+  AddChar.compAddMonoidHom ⟨AddCircle.toCircle, AddCircle.toCircle_zero, AddCircle.toCircle_add⟩ <|
+    ZMod.toAddCircle.comp <| .mulLeft x
+
+@[simp] lemma zmod_intCast (x y : ℤ) : zmod n x y = exp (2 * π * (x * y / n)) := by
+  simp [zmod, ← Int.cast_mul x y, -Int.cast_mul, ZMod.toAddCircle_intCast,
+    AddCircle.toCircle_apply_mk]
+
+@[simp] lemma zmod_zero : zmod n 0 = 1 :=
+  DFunLike.ext _ _ <| by simp [ZMod.intCast_surjective.forall, zmod]
+
+variable {n}
+
+@[simp] lemma zmod_add : ∀ x y : ZMod n, zmod n (x + y) = zmod n x * zmod n y := by
+  simp [DFunLike.ext_iff, ← Int.cast_add, zmod, add_mul, add_div, map_add_eq_mul]
+
+lemma zmod_injective : Injective (zmod n) := by
+  simp_rw [Injective, ZMod.intCast_surjective.forall]
+  rintro x y h
```

Comment:
```suggestion
  intro x y h
```

Reply:
I don't really see the point of learning when to use `intro` when it is strictly weaker than `rintro`. If you ask me, I would be very happy for mathlib to make `intro` notation for `rintro`

Reply:
OK; I don't insist...